docs/spec-files/encryption.md

```diff
@@ -146,6 +146,21 @@ Even though `apu`/`apv` are not mandatory JWE recipients headers, they are requi
 
 A final note about `skid` header: since the 1PU draft [does not require](https://datatracker.ietf.org/doc/html/draft-madden-jose-ecdh-1pu-04#section-2.2.1) this header, authcrypt implementations MUST be able to resolve the sender kid from the `apu` header if `skid` is not set.
 
+#### ECDH-ES key wrapping and common protected headers
+
+When using anoncrypt, any of the valid content encryption algorithms may be used. To meet this requirement, JWE messages MUST use common `epk`, `apv` and `alg` headers for all recipient keys. They MUST be set in the `protected` headers JWE section.
```

Comment:
I'm not sure why "protected" is formatted as code here. Is there actually a section of the output data structure that is tagged in this way? In other places, I've just seen the phrase "protected headers" without any special formatting...

Reply:
> To meet this requirement, JWE messages MUST use common `epk`, `apv` and `alg` headers for all recipient keys. They MUST be set in the `protected` headers JWE section.

This is only true for a single recipient envelopes, which can be set as [compact serialized JWE format](https://datatracker.ietf.org/doc/html/rfc7516#section-7.1) (which supports a single recipient only) or [Flattened JSON serialization](https://datatracker.ietf.org/doc/html/rfc7516#section-7.2.2) format (it also supports a single recipient only). In these formats, the recipient specific headers can be merged into the Protected headers and `recipients` section is skipped.

For multi-recipient envelopes, each recipient must have their own `header` and `encrypted_key` section as per the [JWE serialization formatting rules](https://datatracker.ietf.org/doc/html/rfc7516#:~:text=single%2Drecipient%20case.-,7.2.1,-.%20%20General%20JWE%20JSON). 

I'd probably avoid sharing the same encrypted key (encrypted cek) value with all recipients in the case where recipients represent different entities (ie if the recipients list comprises my different devices keys only, then it's ok to share, but if it represents different users, then it's not ok to share for additional protection). In the end, each recipient must own the private key used in the ECDH key wrapping in order to unwrap the cek. This is why each recipient must have their own `header` section under `recipients` in the general JSON format. 

In the case of a single recipient, then compact or JSON Flattened format can be used. This is the only case I see where `epk`, `apv` and `apu` can be merged into the `protected` headers.

Reply:
@Baha-sk : Thanks for the comments.

We've decided that we will use multi-recipient features to send the same message to multiple devices belonging to a single party (e.g., Bob has 3 agents); however, we will not use multi-recipient features to send the same message to Bob, Carol, and David. The latter doesn't make sense, since Carol and David may have different key types from Bob, and will certainly have different mediators and routes.

Does this alter your concerns in any way?

Reply:
> @Baha-sk : Thanks for the comments.
> 
> We've decided that we will use multi-recipient features to send the same message to multiple devices belonging to a single party (e.g., Bob has 3 agents); however, we will not use multi-recipient features to send the same message to Bob, Carol, and David. The latter doesn't make sense, since Carol and David may have different key types from Bob, and will certainly have different mediators and routes.
> 
> Does this alter your concerns in any way?

Thanks for the clarification @dhh1128 

My only concern is mostly about spec compliance to be honest. 

My understanding from the JWE serialization sepc I linked above is it entails 1 key per recipient regardless of who owns the key. 1 user having 3 device keys = 3 recipients and therefore must use General JSON format.


If I understand correctly, what we're aiming for here is to have a hybrid JWE model: Use a JSON Flattened format with multi recipient keys. I'm not sure it's doable since key Wrapping (and its ECDH key derivation) must be done separately for each key which will generate unique `encrypted_key` entries. How can this list be formatted in a single recipient JSON Flattened format?

Reply:
>We've decided that we will use multi-recipient features to send the same message to multiple devices belonging to a single party (e.g., Bob has 3 agents);

if this means sending separate JWE envelopes for each device, then we're ok. I think the misunderstanding came from sending the same envelope to all three devices at once.

I think we can clarify this by saying that `each device will obtain its unique envelope targeted to its respective key` or something along this line.

Reply:
TODO: Need to remove the tick marks around protected as it's not a literal value in the JWE.

Reply:
Nevermind. keep ticks, remove 'headers' from 'protected' _headers_ in both auth and anoncrypt sections.